sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/SchedulerConfig.java

```diff
@@ -386,6 +409,34 @@ public String getDcosSpace() {
     return Optional.ofNullable(envStore.getOptional(SERVICE_REGION_ENV, null));
   }
 
+  public Optional<String> getServiceNamespace() {
+    boolean enforceRole = envStore.getOptionalBoolean(MARATHON_APP_ENFORCE_GROUP_ROLE_ENV, false);
+    if (enforceRole) {
+      // If enforceRole is set, we *must* use the value mandated by MESOS_ALLOCATION_ROLE
+      return Optional.of(envStore.getRequired(MESOS_ALLOCATION_ROLE_ENV));
+    } else {
+      // We're not in a Marathon group with enforced roles, see if the user has specified a preferred role.
+      String preferredServiceRole = envStore.getOptional(MESOS_ALLOCATION_ROLE_ENV, null);
+
+      // If the user specifies an invalid role, ie `service.role=prod` when the
+      // service name is `dev/foo-service` and is under the group `dev`.
+      // Marathon validation resets the role to the default of `slave_public`.
+      // We cannot launch all pods under `slave_public` as it is bad form.
+      if (preferredServiceRole != null && !DEFAULT_SCHEDULER_ROLE.equals(preferredServiceRole)) {
+        // Here enforceRole is not set and we have valid preferred service role, we can use this.
+        return Optional.of(preferredServiceRole);
+      } else {
+        // Here enforceRole is not set and we have an invalid preferred service role we can't use.
+        // Revert to legacy semantics.
+        return Optional.empty();
```

Comment:
Should we revert to legacy semantics or should we fail loudly ?! 

I think of it as a reserved keyword in mustache template and thus if a user mentions `role` field (may be with some other intentions) then it should fail. User should not specify invalid roles

Reply:
It's Marathon that does the validation for us here, essentially either the specified role is is correct or we revert to `slave_public`. So for now I don't think there's a need to fail loudly. If at some point in the future, `slave_public` is deprecated (or something along those lines), we'll fail loudly.